services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/events.py

```diff
@@ -503,114 +506,133 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
+                        "%s\n%s. Will continue service removal!"
                     ),
                     scheduler_data.service_name,
                     f"{e}",
                 )
-                # ensure dynamic-sidecar does not get removed
-                # user data can be manually saved and manual
-                # cleanup of the dynamic-sidecar is required
-                # TODO: ANE: maybe have a mechanism stop the dynamic sidecar
-                # and make the director warn about hanging sidecars?
-                raise e
-
-        # remove the 2 services
-        await remove_dynamic_sidecar_stack(
-            node_uuid=scheduler_data.node_uuid,
-            dynamic_sidecar_settings=dynamic_sidecar_settings,
-        )
-        # remove network
-        await remove_dynamic_sidecar_network(
-            scheduler_data.dynamic_sidecar_network_name
-        )
+            app_settings: AppSettings = app.state.settings
+            dynamic_sidecar_settings: DynamicSidecarSettings = (
+                app_settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR
+            )
+
+            # only try to save the status if :
+            # - the dynamic-sidecar has finished booting correctly
+            # - it is requested to save the state
+            if (
+                scheduler_data.dynamic_sidecar.service_removal_state.can_save
+                and await are_all_services_present(
+                    node_uuid=scheduler_data.node_uuid,
+                    dynamic_sidecar_settings=app.state.settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR,
+                )
+            ):
+                dynamic_sidecar_client = get_dynamic_sidecar_client(app)
+                dynamic_sidecar_endpoint = scheduler_data.dynamic_sidecar.endpoint
 
-        # NOTE: for future attempts, volumes cannot be cleaned up
-        # since they are local to the node.
-        # That's why anonymous volumes are used!
+                logger.info(
+                    "Calling into dynamic-sidecar to save state and pushing data to nodeports"
+                )
+                try:
+                    tasks = [
+                        dynamic_sidecar_client.service_push_output_ports(
+                            dynamic_sidecar_endpoint,
+                        )
+                    ]
+                    # When enabled no longer uploads state via nodeports
+                    # S3 is used to store state paths
+                    if not app_settings.DIRECTOR_V2_DEV_FEATURES_ENABLED:
+                        tasks.append(
+                            dynamic_sidecar_client.service_save_state(
+                                dynamic_sidecar_endpoint,
+                            )
+                        )
 
-        logger.debug(
-            "Removed dynamic-sidecar created services for '%s'",
-            scheduler_data.service_name,
-        )
+                    try:
+                        await logged_gather(*tasks)
```

Comment:
be careful with these gather calls. How many tasks are called here? I started using the max_concurrency argument..

Reply:
Only 2 tasks are gathered here at most.

Reply:
MINOR: if that is so, perhaps assert it? (defensive programming?)

Reply:
Let me put a hard limit on it to be sure.